### PR TITLE
Adding Encoding support

### PR DESCRIPTION
**Pull request by lablnet**
https://github.com/zestframework/Zest_Framework/issues/247

### CODE REVIEW

File: src/Contracts/Data/Str.php [hunk 1]
```diff
@@ -21,19 +21,20 @@ interface Str
...
-     * @param string $str String to be evaluated.
+     * @param string $str      String to be evaluated.
+     * @param string $encoding Valid encoding.
      *
      * @since 3.0.0
      *
      * @return string
      */
-    public static function reverse(string $str) :string;
+    public static function reverse(string $str, $encoding = null) :string;
```
Comment (peter279k) [inline src/Contracts/Data/Str.php:31]: It looks like the `$encoding` type is `string`, and I think the `$encoding` default value should be empty.

If we set the `$encoding` is `null`,the `$encoding` `@param` annotation should not only `string`.

File: src/Contracts/Data/Str.php [hunk 2]
```diff
@@ -44,11 +45,12 @@ public static function concat($g, ...$str);
...
-     * @param string $str String to be counted.
+     * @param string $str      String to be counted.
+     * @param string $encoding Valid encoding.
      *
      * @since 3.0.0
      *
      * @return int
      */
-    public static function count(string $str);
+    public static function count(string $str, $encoding = null);
```
Comment (peter279k) [inline src/Contracts/Data/Str.php:55]: The situation is same as previous comment.

File: src/Data/Str.php [hunk 3]
```diff
@@ -21,23 +21,39 @@
...
      *
-     * @param string $str String to be evaluated.
+     * @param string $encoding Valid encoding.
      *
      * @since 3.0.0
      *
      * @return string
      */
-    public static function reverse(string $str) :string
+    private static function encoding($encoding = null) :string
```
Comment (peter279k) [inline src/Data/Str.php:32]: This situation I mention on previous comment thread..

File: src/Data/Str.php [hunk 4]
```diff
@@ -21,23 +21,39 @@
...
     {
-        if (function_exists('strrev')) {
-            return strrev($str);
-        }
+        return $encoding ?: \mb_internal_encoding();
```
Comment (peter279k) [inline src/Data/Str.php:34]: The `mb_internal_encoding` function is available on `php-mbstring` extension.

It should check the `mbstring` is available during composer installation.

File: src/Contracts/Data/Str.php (hunk 1 above)
Comment (lablnet) [reply src/Contracts/Data/Str.php:31]: SO it should be mixed?

File: src/Data/Str.php [hunk 5]
```diff
@@ -61,16 +77,17 @@ public static function concat($g, ...$str)
...
-     * @param string $str String to be counted.
+     * @param string $str      String to be counted.
+     * @param string $encoding Valid encoding.
      *
      * @since 3.0.0
      *
      * @return int
      */
-    public static function count(string $str)
+    public static function count(string $str, $encoding = null)
```
Comment (peter279k) [inline src/Data/Str.php:87]: This situation I've mentioned on previous comment.

File: src/Data/Str.php (hunk 4 above)
Comment (lablnet) [reply src/Data/Str.php:34]: If not available we have to install it?

I think you may send PR, in your available time

File: src/Contracts/Data/Str.php (hunk 1 above)
Comment (peter279k) [reply src/Contracts/Data/Str.php:31]: The `param` should be `$encoding|null`.

And function name should be `public static function reverse(string $str, ?$encoding) :string;`.

This syntax is available on `php-7.1+`.

Comment (peter279k) [reply src/Contracts/Data/Str.php:31]: The reference is available on [PHPDoc](https://docs.phpdoc.org/guides/types.html#multiple-types-combined).

Comment (lablnet) [reply src/Contracts/Data/Str.php:31]: Ok thanks you

so this syntax `?$encoding` tell it can null means to say optional right?

Comment (peter279k) [reply src/Contracts/Data/Str.php:31]: Yes. That's right.

**Pull request merged**